agent/pkg/agent/agent.go

```diff
@@ -0,0 +1,692 @@
+	}
+
+	// Extract message length from first 2 bytes
+	// - First byte shifted left 8 bits + second byte
+	// - Creates 16-bit length prefix
+	messageLen := int(payload[0])<<8 | int(payload[1])
+	if messageLen == 0 || len(payload) < messageLen+2 {
+		return nil, fmt.Errorf("invalid DNS over TCP payload length")
+	}
+
+	// We attempt to decode the DNS over TCP payload
+	// The only way we can accept the request is if the DNS query is contained within a single TCP packet payload
+	dns := &layers.DNS{}
+	err := dns.DecodeFromBytes(payload[2:messageLen+2], gopacket.NilDecodeFeedback)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode DNS over TCP payload: %w", err)
+	}
+
+	if len(dns.Questions) == 0 {
+		return nil, fmt.Errorf("no DNS questions in payload")
+	}
+
+	return dns, nil
+}
+
+func (a *Agent) processDNSOverTCPPayload(payload []byte, pkt PacketInfo) uint8 {
+	dns, err := extractDNSFromTCPPayload(payload)
+	if err != nil {
+		fmt.Printf("Failed to extract DNS from TCP payload: %v\n", err)
+		return DROP_REQUEST
+	}
+	return a.processDNSLayer(dns, pkt)
+}
+
+func (a *Agent) processDNSOverTCPPacket(packet gopacket.Packet) uint8 {
+	tcpLayer := packet.Layer(layers.LayerTypeTCP)
+	tcp, _ := tcpLayer.(*layers.TCP)
+	dstPort, srcPort, payload := tcp.DstPort, tcp.SrcPort, tcp.Payload
+
+	pkt := a.extractPacketInfo(packet)
+	fmt.Printf("Processing DNS-over-TCP packet: %s:%s -> %s:%s\n",
+		pkt.SrcIP, pkt.SrcPort, pkt.DstIP, pkt.DstPort)
+
+	if len(payload) == 0 {
+		// We only accept DNS over TCP packets with no payload since they are only used for initiating a connection
+		fmt.Printf("Accepting TCP handshake packet: %s:%s -> %s:%s\n",
+			pkt.SrcIP, pkt.SrcPort, pkt.DstIP, pkt.DstPort)
+		return ACCEPT_REQUEST
+	}
+
+	// Extract domain from payload for logging purposes
+	domain := "unknown"
+	dns, err := extractDNSFromTCPPayload(payload)
+	if err == nil && len(dns.Questions) > 0 {
+		domain = string(dns.Questions[0].Name)
+	} else {
+		fmt.Printf("failed to extract dns from TCP payload: %v\n", err)
+	}
+
+	// Validate DNS server IP
+	if dstPort == DNS_PORT {
+		if !a.allowedDNSServers[pkt.DstIP] {
+			fmt.Printf("%s -> Blocked DNS-over-TCP Query. Untrusted DNS server %s\n", domain, pkt.DstIP)
+			a.addConnectionLog(pkt, "blocked", "TCP-DNS", domain, "untrusted-dns-server")
+			return DROP_REQUEST
+		}
+	}
+
+	if dstPort != DNS_PORT && srcPort != DNS_PORT {
+		fmt.Println("Warning: Destination and source port are not DNS ports. Dropping request")
+		return DROP_REQUEST
+	}
+
+	// if we are not blocking DNS queries, just accept the query request
+	if !a.blockDNS && dstPort == DNS_PORT {
+		return ACCEPT_REQUEST
+	}
+
+	// Now we have a payload in the TCP packet, we need to make sure it is a valid DNS over TCP payload and the DNS query is for a known domain. We don't want to exfiltrate data over DNS over TCP
+	return a.processDNSOverTCPPayload(payload, pkt)
+
+}
+
+func (a *Agent) processNonDNSPacket(packet gopacket.Packet) uint8 {
```

Comment:
I know this code/logic predates this change, but this is a note to ourselves for the future. To make things clearer, regardless of whether we are in audit mode or block mode, each decision should return a single, unified result that includes all relevant information.

That means:

- No inline calls to addConnectionLog inside sub-functions.
- A single call to addConnectionLog from processPacket.
- Each sub-process function should return a complete decision object, including a deny decision even when running in audit mode.

Then, processHandler applies the final decision:

- In block mode, it enforces the decision returned by the sub-functions.
- In audit mode, it bypasses enforcement but still records the decision.

Reply:
Yeah, there are clearly more opportunities to clean up the code. I just need to stop the rabbit hole for now :sweat_smile: